snapcraft/internal/lxd.py

```diff
@@ -204,18 +218,83 @@ def _ensure_container(self):
+
     def _ensure_mount(self, destination, source):
         logger.info('Mounting {} into container'.format(source))
+        remote, container_name = self._container_name.split(':')
+        if remote != 'local':
+            self._remote_mount(destination, source)
+        else:
+            devices = self._get_container_status()['devices']
+            if destination not in devices:
+                check_call([
+                    'lxc', 'config', 'device', 'add', self._container_name,
+                    destination, 'disk', 'source={}'.format(source),
+                    'path=/{}'.format(destination)])
+
+    def _remote_mount(self, destination, source):
+        # Remove project folder in case it was used "locally" before
         devices = self._get_container_status()['devices']
-        if destination not in devices:
+        if destination in devices:
             check_call([
-                'lxc', 'config', 'device', 'add', self._container_name,
+                'lxc', 'config', 'device', 'remove', self._container_name,
                 destination, 'disk', 'source={}'.format(source),
                 'path=/{}'.format(destination)])
 
+        # Generate an SSH key and add it to the container's known keys
+        keyfile = 'id_{}'.format(self._container_name)
+        if not os.path.exists(keyfile):
+            check_call(['ssh-keygen', '-o', '-N', '', '-f', keyfile],
+                       stdout=os.devnull)
+        self._container_run(['mkdir', '-p', '/root/.ssh'])
+        self._container_run(['chmod', '700', '/root/.ssh'])
+        self._container_run(['tee', '-a',
+                             '/root/.ssh/authorized_keys'],
+                            stdin=open('{}.pub'.format(keyfile), 'r'))
+        self._container_run(['chmod', '600',
+                             '/root/.ssh/authorized_keys'])
+
+        # Use sshfs in slave mode inside SSH to reverse mount destination
+        self._install_packages(['sshfs'])
+        self._container_run(['mkdir', '-p', '/{}'.format(destination)])
+        self._container_run(['mkdir', '-p', source])
+        ssh_address = self._get_container_address()
+        logger.info('Connecting via SSH to {}'.format(ssh_address))
+        stdin1, stdout1 = os.pipe()
+        stdin2, stdout2 = os.pipe()
+        try:
+            self._host_run(['/usr/lib/sftp-server'],
+                           stdin=stdin1, stdout=stdout2)
+        except CalledProcessError as e:
+            # XXX: This needs to be extended once we support other distros
+            raise SnapcraftEnvironmentError(
+                'sftp-server could not be run.\n'
+                'On Debian, Ubuntu and derivatives, the package '
+                'openssh-sftp-server needs to be installed.'
+                ) from e
+        self._host_run(['ssh', '-C', '-F', '/dev/null',
+                        '-o', 'IdentityFile={}'.format(keyfile),
+                        '-o', 'StrictHostKeyChecking=no',
```

Comment:
I didn't want to make this a running target, what with IP changes and such - since we're in charge I figured it's safe to ignore that check, we're always SSHing into the container and using the key we generated.